### PR TITLE
Add notebook on fine tuning for function calling

### PR DESCRIPTION
**Pull request by jhills20**
## Summary

Adding fine tuning for function calling recipe. This generates synthetic data to improve the performance of function calling for a 'drone co-pilot ai'. The notebook shows how an untuned 3.5 Turbo fails to reject certain requests that the drone cannot do, and improperly calls functions. Then, the notebook creates invocations for every function, generates synthetic prompts that would lead to that invocation, and fine-tunes the model. The result is an improved model that properly rejects those questions.

## Motivation

There is no notebook now on fine tuning for function calling, and this is a big customer need and ask.

---

## For new content

When contributing new content, read through our [contribution guidelines](/CONTRIBUTING.md), and mark the following action items as completed:

- [x] I have added a new entry in [registry.yaml](/registry.yaml) so that my content renders on the cookbook website.
- [x] I have conducted a self-review of my content based on the [c](/CONTRIBUTING.md#rubric):
  - [x] Relevance: This content is related to building with OpenAI technlogies and is useful to others.
  - [x] Uniqueness: I have searched for related examples in the OpenAI Cookbook, and verified that my content offers new insights or unique information compared to existing documentation.
  - [x] Spelling and Grammar: I have checked for spelling or grammatical mistakes.
  - [x] Clarity: I have done a final read-through and verified that my submission is well-organized and easy to understand.
  - [x] Correctness: The information I include is correct and all of my code executes successfully.
  - [x] Completeness: I have explained everything fully, including all necessary references and citations.

We will rate each of these areas on a scale from 1 to 4, and will only accept contributions that score 3 or higher on all areas. Refer to our [contributions guidelines](/CONTRIBUTING.md) for more details.

### CODE REVIEW

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 1]
```diff
@@ -0,0 +1,1248 @@
...
+    ") -> str:\n",
+    "    completion = openai.ChatCompletion.create(\n",
+    "        model=model,\n",
+    "        messages=messages,\n",
+    "        max_tokens=max_tokens,\n",
+    "        temperature=temperature,\n",
+    "        stop=stop,\n",
+    "    )\n",
+    "    return completion.choices[0].message\n",
+    "\n"
```
Comment (ibigio) [inline examples/Fine_tuning_for_function_calling.ipynb:144]: Any reason not to just have a single one used both for functions and completions?

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 2]
```diff
@@ -0,0 +1,1248 @@
...
+   "source": [
+    "### Helper functions"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We want to generate every invocation of every function, so that we have\n",
+    "full coverage of all potential invocations to create synthetic data for. Then, we will use GPT-4 to come up with prompts that would call each invocation, and we will use that prompt - function invocation pair as training data."
```
Comment (ibigio) [inline examples/Fine_tuning_for_function_calling.ipynb:595]: Might want to motivate this – e.g. To perform effective fine-tuning we need correctly labeled data. We could manually come up with examples and label the data... or we could generate synthetic data with the help of GPT4. One way to generate a comprehensive dataset it to...

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 3]
```diff
@@ -0,0 +1,1248 @@
...
+    "placeholder_string = 'fill_in_string'\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def generate_permutations(params: Dict[str, Dict[str, Any]]) -> Generator[Dict[str, Any], None, None]:\n",
```
Comment (ibigio) [inline examples/Fine_tuning_for_function_calling.ipynb:622]: Might want to document/explain this a bit more.

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 4]
```diff
@@ -0,0 +1,1248 @@
...
+   "outputs": [],
+   "source": [
+    "input_objects = []\n",
+    "all_but_reject = [f for f in function_list if f.get('name') != 'reject_request']\n",
+    "\n",
+    "for function in all_but_reject:\n",
+    "    func_name = function[\"name\"]\n",
+    "    params = function[\"parameters\"]\n",
+    "    for arguments in generate_permutations(params):\n",
+    "      #  if arguments.values().([placeholder_int,placeholder_string]):\n",
```
Comment (ibigio) [inline examples/Fine_tuning_for_function_calling.ipynb:751]: 👀

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 5]
```diff
@@ -0,0 +1,1248 @@
...
+   "execution_count": 21,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def create_commands(invocation_list):\n",
+    "    example_list = []\n",
+    "    for i, invocation in enumerate(invocation_list):\n",
+    "        print(f'\\033[34m{np.round(100*i/len(invocation_list),1)}% complete\\033[0m')\n",
+    "        print(invocation)\n",
+    "    # Format the prompt with the invocation string\n",
```
Comment (ibigio) [inline examples/Fine_tuning_for_function_calling.ipynb:786]: nit: fix indentation

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 6]
```diff
@@ -0,0 +1,1248 @@
...
+    "- **Assessing baseline function calling performance:** Evaluating an out-of-the-box 3.5 Turbo model on our given function (let's assume that for latency + cost reasons we cannot use GPT-4 for a drone copilot)\n",
+    "- **Generating synthetic data:** Using GPT-4 to create 'golden' set of prompts and function invocations to use as training data\n",
+    "- **Fine-tuning**: Running the fine tuning job, and evaluating the fine-tuned model\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note: *This notebook provides an example of how to create synthetic training data for fine tuning for function calling given just a list of functions. While real-world production test evals are preferable, this method produces strong results and can (and should) be used in conjuction with real-world training data.*"
```
Comment (ibigio) [inline examples/Fine_tuning_for_function_calling.ipynb:48]: (and should) <- this might vary, could be cases it doesn't help, so we might wanna leave it at just "can"

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 7]
```diff
@@ -0,0 +1,1248 @@
...
+   "metadata": {},
+   "source": [
+    "### Overview\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This notebook contains three sections\n",
```
Comment (ibigio) [inline examples/Fine_tuning_for_function_calling.ipynb:38]: Before we even get here it could be good to motivate the entire notebook – _why_ should someone do fine-tuning with function calling? Why are we doing these steps? (maybe even how we got here?)

**Review by ibigio — COMMENTED** [review]
This is going to be such an awesome notebook to share with people – love it!

File: examples/Fine_tuning_for_function_calling.ipynb (hunk 1 above)
Comment (jhills20) [reply examples/Fine_tuning_for_function_calling.ipynb:144]: For whatever reason using using the function calling with functions=None was failing, may have just been a me issue

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 8]
```diff
@@ -0,0 +1,1274 @@
...
+    "    print(completion.content)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we run into some problems.\n",
+    "The model here should reject all of these requests, as they are impossible given the functions, however instead the model calls functions that are somewhat related to the request, but incorrect. The model sets the camera to video when asked to begin 'live streaming to social media', and changes the LED's to blue when asked to 'change the paint color'... \\\n",
+    "Even with prompt engineering (in the prompt we say to decline requests when there is any ambiguity or doubt), the function calling is not working well. So this is a good opportuntiy for fine tuning for function calling."
```
Comment (simonpfish) [inline examples/Fine_tuning_for_function_calling.ipynb:587]: I'm not fully convinced by this, it feels to me like we just need better prompt engineering here. But maybe for the purpose of the example we can just say something like "in this simple case, these issues might be solvable through prompt engineering, but for the purpose of the example we will demonstrate how fine-tuning can be used instead"

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 9]
```diff
@@ -0,0 +1,1274 @@
...
+    "We want to generate every invocation of every function, so that we have\n",
+    "full coverage of all potential invocations to create synthetic data for. Then, we will use GPT-4 to come up with prompts that would call each invocation, and we will use that prompt - function invocation pair as training data."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Generating every invocation for a function with fixed enums is more simple, but for a function such as\n",
+    "  `control_gimbal` we need to set the `tilt` and `pan` values which are integers, so to generate those synthetic invocations we will first set a place holder, and then later use GPT-4 to come up with reasonable values."
```
Comment (simonpfish) [inline examples/Fine_tuning_for_function_calling.ipynb:617]:
```suggestion
    "  `control_gimbal` we need to set the `tilt` and `pan` integer values, so to generate those synthetic invocations we will first set a placeholder, and then later use GPT-4 to come up with reasonable values."
```

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 10]
```diff
@@ -0,0 +1,1274 @@
...
+    "  for required_perm in required_permutations:\n",
+    "      yield from generate_optional_permutations(params, required_perm)\n",
+    "\n",
+    "def generate_required_permutations(params: Dict[str, Dict[str, Any]], required_fields: List[str]) -> List[Dict[str, Any]]:\n",
+    "  required_values = [get_possible_values(params, field) for field in required_fields]\n",
+    "  return [dict(zip(required_fields, values)) for values in itertools.product(*required_values)]\n",
+    "\n",
+    "def generate_optional_permutations(params: Dict[str, Dict[str, Any]], base_perm: Dict[str, Any]) -> Generator[Dict[str, Any], None, None]:\n",
+    "  optional_fields = set(params['properties']) - set(base_perm)\n",
+    "  for field_subset in itertools.chain.from_iterable(itertools.combinations(optional_fields, r) for r in range(len(optional_fields)+1)):\n",
```
Comment (simonpfish) [inline examples/Fine_tuning_for_function_calling.ipynb:658]: This code is a bit hard to parse, let's maybe add some comments and some spacing. An autoformatter might help. You can also split things into functions. I bet GPT-4 has some ideas for how to make it more readable.

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 11]
```diff
@@ -0,0 +1,1274 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 44,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Already a dict\n",
```
Comment (simonpfish) [inline examples/Fine_tuning_for_function_calling.ipynb:969]: These logs seem unnecessary

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 12]
```diff
@@ -0,0 +1,1274 @@
...
+    "Now, back to the rejection function. Let's generate some prompts that are *nearly* possible, but should result in the `decline_request` function being called."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 45,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# messages.append({\"role\": \"user\", \"content\": f\"Come up with a list, returned in brackets [], of 20 prompts that would call functions related to the drone but are not quite possible with the given functions: {function_list}\"})\n",
```
Comment (simonpfish) [inline examples/Fine_tuning_for_function_calling.ipynb:1057]: Should we remove these lines?

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 13]
```diff
@@ -0,0 +1,1274 @@
...
+   "metadata": {},
+   "source": [
+    "# Fine tuning"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Finally, we can kick off the FT'ing job"
```
Comment (simonpfish) [inline examples/Fine_tuning_for_function_calling.ipynb:1132]:
```suggestion
    "Finally, we can kick off the fine-tuning job"
```

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 14]
```diff
@@ -0,0 +1,1274 @@
...
+   "metadata": {},
+   "source": [
+    "# Evaluations"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Great! We trainied a fine-tuned model for function calling. Let's see how it does on our evaluation set for prompts that the drone assistant\n",
```
Comment (simonpfish) [inline examples/Fine_tuning_for_function_calling.ipynb:1175]:
```suggestion
    "Great! We trained a fine-tuned model for function calling. Let's see how it does on our evaluation set for prompts that the drone assistant\n",
```

File: examples/Fine_tuning_for_function_calling.ipynb [hunk 15]
```diff
@@ -0,0 +1,1274 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Fine tuning for function calling (with synthetic data)"
```
Comment (simonpfish) [inline examples/Fine_tuning_for_function_calling.ipynb:7]: Suggestion: 
```suggestion
    "# Fine tuning on synthetic function-calling data"
```

**Review by simonpfish — CHANGES_REQUESTED** [review]
This is a great notebook! People will find this very valuable. There is room for improvement in clarity, correctness, and grammar. I left some comments throughout the notebook.

| Criteria     | Description                                                                                         | Score | Explanation                                                                                                      |
|--------------|-----------------------------------------------------------------------------------------------------|-------|------------------------------------------------------------------------------------------------------------------|
| Relevance    | Is the content related to building with OpenAI technologies? Is it useful to others?                | 4     | Your submission is highly relevant and provides useful information for individuals working with OpenAI tech.      |
| Uniqueness   | Does the content offer new insights or unique information compared to existing documentation?       | 4     | The content presents completely unique perspectives and insights not covered in existing documentation.            |
| Clarity      | Is the language easy to understand? Are things well-explained? Is the title clear?                  | 3     | The language is clear, but there may be some sections where the structure or explanations could be improved.       |
| Correctness  | Are the facts, code snippets, and examples correct and reliable? Does everything execute correctly? | 3     | The factual information is accurate; however, there might be minor errors or improvements needed for code snippets.|
| Completeness | Is the content thorough and detailed? Are there things that weren’t explained fully?                | 4     | The content is comprehensive and detailed, covering all necessary aspects of the topic.                           |
| Grammar      | Are there grammatical or spelling errors present?                                                   | 3     | The submission has correct grammar with a few typos present.                                                      |

**Review by jhills20 — COMMENTED** [review]
All changes should be made

**Comment by simonpfish** [comment]
Looks great! 🎉
Merging
